base/nullable.jl

```diff
@@ -78,3 +78,30 @@ function hash(x::Nullable, h::UInt)
         return hash(x.value, h + nullablehash_seed)
     end
 end
+
+# indexing is either without index, or with 1 as index
+function getindex(x::Nullable)
+    @boundscheck x.isnull && throw(NullException())
+    x.value
+end
+function getindex(x::Nullable, i::Integer)
+    @boundscheck (x.isnull | (i ≠ 1)) && throw(BoundsError(i, x))
+    x.value
+end
+
+# higher-order functions
+function filter{T}(p, x::Nullable{T})
+    if x.isnull || p(x.value)
+        x
+    else
+        Nullable{T}()
+    end
+end
+
+function map{T}(f, x::Nullable{T})
+    if x.isnull
+        Nullable{Union{}}()  # Union{} for consistency with arrays
```

Comment:
Inference is now used to decide what's the best type:
```julia
julia> map(exp, Int[])
0-element Array{Float64,1}

```

So use the same strategy as `lift` via `Core.Inference.return_type`.